rust/core-lib/src/editor.rs

```diff
@@ -732,44 +792,40 @@ impl<W: Write + Send + 'static> Editor<W> {
-        let (start, middle) = if end_opt.is_none() && start_opt.is_some() {
-            // if at the very end, swap previous TWO characters (instead of ONE)
-            let middle = start_opt.unwrap();
-            let start = self.text.prev_grapheme_offset(middle).unwrap_or(middle);
-            (start, middle)
-        } else {
-            (start_opt.unwrap_or(self.view.sel_end), self.view.sel_end)
-        };
-
-        let interval = Interval::new_closed_open(start, end);
-        let swapped = self.text.slice_to_string(middle, end) +
-                      &self.text.slice_to_string(start, middle);
-        self.add_simple_edit(interval, Rope::from(swapped));
+        let mut builder = delta::Builder::new(self.text.len());
+        let mut last = 0;
+        for region in self.view.sel_regions() {
+            if region.is_caret() {
+                let middle = region.end;
+                let start = self.text.prev_grapheme_offset(middle).unwrap_or(0);
+                // Note: this matches Sublime's behavior. Cocoa would swap last
+                // two characters of line if at end of line.
+                if let Some(end) = self.text.next_grapheme_offset(middle) {
+                    if start >= last {
+                        let interval = Interval::new_closed_open(start, end);
+                        let swapped = self.text.slice_to_string(middle, end) +
+                                      &self.text.slice_to_string(start, middle);
+                        builder.replace(interval, Rope::from(swapped));
+                        last = end;
+                    }
+                }
+            }
+            // TODO: handle else case by rotating non-caret regions.
+        }
+        if !builder.is_empty() {
+            self.this_edit_type = EditType::Other;
+            self.add_delta(builder.build());
+        }
     }
 
     fn delete_to_end_of_paragraph(&mut self) {
-        let current = self.view.sel_max();
-        let offset = self.end_of_paragraph_offset();
-        let mut val = String::from("");
-
-        if current != offset {
-            val = self.text.slice_to_string(current, offset);
-            let del_interval = Interval::new_closed_open(current, offset);
-            self.add_simple_edit(del_interval, Rope::from(""));
-        } else if let Some(grapheme_offset) = self.text.next_grapheme_offset(self.view.sel_end) {
-            val = self.text.slice_to_string(current, grapheme_offset);
-            let del_interval = Interval::new_closed_open(current, grapheme_offset);
-            self.add_simple_edit(del_interval, Rope::from(""));
-        }
-
-        self.doc_ctx.set_kill_ring(Rope::from(val));
+        self.delete_by_movement(Movement::EndOfParagraphKill, true);
     }
 
     fn yank(&mut self) {
+        // TODO: if there are multiple cursors and the number of newlines
+        // is one less than the number of cursors, split and distribute one
+        // line per cursor.
```

Comment:
So that's an interesting question. One compelling reason to use a string as the intermediate representation rather than something with more structure is that it can use the plain text type, and interoperate with other things. My sense is that the cases where you need the structure are going to be pretty marginal. It's worth considering for the long term though.